api/search.go

```diff
@@ -113,49 +114,71 @@ func SearchHandlerFunc(queryBuilder QueryBuilder, elasticSearchClient DpElasticS
-			log.Error(ctx, "elastic search returned invalid JSON for search query", errors.New("elastic search returned invalid JSON for search query"))
-			http.Error(w, "Failed to process search query", http.StatusInternalServerError)
-			return
+		var resDataChan = make(chan []byte)
+		var resCountChan = make(chan []byte)
+		var responseSearchData []byte
+		go func() {
+			defer close(resCountChan)
+			processCountQuery(ctx, elasticSearchClient, queryBuilder, sanitisedQuery, resCountChan)
+		}()
+		go func() {
+			defer close(resDataChan)
+			processSearchQuery(ctx, elasticSearchClient, queryBuilder, sanitisedQuery, typesParam, sort, topicSlice, limit, offset, resDataChan)
+		}()
+
+		for responseData := range resDataChan {
+			if responseData == nil {
+				log.Error(ctx, "call to elastic multisearch api failed", errors.New("nil response data"))
+				http.Error(w, "call to elastic multisearch api failed", http.StatusInternalServerError)
+				return
+			}
+			if !paramGetBool(params, "raw", false) {
+				responseData, err = transformer.TransformSearchResponse(ctx, responseData, q, highlight)
+				if err != nil {
+					log.Error(ctx, "transformation of response data failed", err)
+					http.Error(w, "failed to transform search result", http.StatusInternalServerError)
+					return
+				}
+			}
+			responseSearchData = responseData
 		}
 
-		if !paramGetBool(params, "raw", false) {
-			responseData, err = transformer.TransformSearchResponse(ctx, responseData, q, highlight)
-			if err != nil {
-				log.Error(ctx, "transformation of response data failed", err)
-				http.Error(w, "Failed to transform search result", http.StatusInternalServerError)
+		for responseCountData := range resCountChan {
+			if responseCountData == nil {
+				log.Error(ctx, "call to elastic count api failed due to", errors.New("nil response data"))
+				http.Error(w, "call to elastic count api failed due to", http.StatusInternalServerError)
 				return
 			}
+			if !paramGetBool(params, "raw", false) {
+				count, CountAPIErr := transformer.TransformCountResponse(ctx, responseCountData)
+				if CountAPIErr != nil {
+					log.Error(ctx, "transformation of response count data failed", CountAPIErr)
+					http.Error(w, "failed to transform count result", http.StatusInternalServerError)
+					return
+				}
+				//TODO: This needs to be refactored as it involves multiple marshal and unmarshal code. So basically the
+				// transformSearchResponse function can return an interface that would satisfy both legacy search response and
+				// new search response instead of bytes. So here we just have to add the count instead of unmarshalling the bytes
+				// and adding the count and marshalling it again. Will be done in a separate pr very soon.
+				var esSearchResponse models.SearchResponse
+				if SearchRespErr := json.Unmarshal(responseSearchData, &esSearchResponse); SearchRespErr != nil {
+					log.Error(ctx, "failed to un marshal the essearchResponse data due to", SearchRespErr)
```

Comment:
Better to be explicit in log message

```suggestion
					log.Error(ctx, "failed to unmarshal the elasticsearch response data due to", SearchRespErr)
```

Reply:
Fixed in update 1.